### PR TITLE
Added Linearization and correlation functions

### PR DESCRIPTION
**Pull request by Ve2mrx**
--- Impact to existing code: NONE
Code added as new, optional functions.

Created a function to read the raw 32 bits of the MAX31855 for further processing;
Created a function to convert the raw bits and return the thermocouple temperature in Celsius;
Created a function to convert the raw bits and return the Internal reference temperature (in Celsius);
Created a function that takes the two temperatures and correct linearity errors of the measurement for much better accuracy.

Integrated the fix to the mask bits from the pull-request #13 :
[Fix mask for sign-extending celsius value #13]( #https://github.com/adafruit/Adafruit-MAX31855-library/pull/13)

Simplified CJT linearization (faster processing, same accuracy) as described in the post:
https://forums.adafruit.com/viewtopic.php?f=19&t=32086&start=75#p579783

> Re: Max 31855 Reading High Temperatures
> 
> by zike on Sun Apr 23, 2017 7:24 pm
> Many thanks for this! Wonderfully executed and exquisitely documented. I am happily using this routine with the MAX31855 breakout and a datalogging shield on an UNO to read, log, and safety-interlock my 1250C heat treating furnace.
> 
> I thought it may be helpful to point out that the first part of the code, inverting and linearizing the cold junction comp, is not strictly necessary. The fixed Seebeck coefficient used in the 31855 for CJC (40.73 uV/C) is chosen to afford less than 0.1C integral linearity error _over the service temperature range of the chip_ . By comparison, the ADC resolution corresponds to 0.25C, and the internal thermometer's absolute error can be ±2C to start with.
> 
> We therefore suffer no loss of accuracy by substituting the following line:
> internalVoltage = internalTemp * 0.04073 ; // (mV) MAX38155 comp scale is 40.73 uV/C
> or the entirety of Step 3. This saves considerable memory and execution time!
> 
> (The pseudo-floating point exp() and pow() are huge time sucks for a mere UNO).
> 
> I hope this can help your code fit into other apps (like mine) where RAM and loop time are on the bare edge.
> 
> Thanks again for all your hard work,
> 
> Mike

Thanks!
Martin

### CODE REVIEW

**Comment by rswift** [comment]
Most likely this PR will languish like the others on this codebase, not entirely clear why seems to be no appetite to merge from the community, especially given the positive commentary on the Adafruit forum... Good work though @Ve2mrx :)

**Comment by Ve2mrx** [comment]
Hi Robert!
  The main reason I do it is that I find it useful to look in others PRs for enhancements and ideas. It may never be merged, but the "add-on" is ready for everyone to integrate in their own projects.

  Also, as I once read, backward code compatibility is critical. I think that it is better to create new functions than modifying existing ones, that way I change nothing of the existing compatibility while adding features. 

  Of course, human resources are limited too, and they are prioritizing. I think PRs for Seesaw and Python stuff has a good chance to be reviewed and maybe merged! I had a Seesaw issue the other day, and it got fixed at night during the weekend! 

  I think the will is there, but not the means!

Have a great day!
Martin

**Comment by tablatronix** [comment]
This might be nice to also have getters for all reads, as this library does a new read for all functions which is nice and annoying at the same time.

`read` - update bitraw cache

getters for bit raw cache var
`getError`
`getInternal` 
`getCelcius`
`getFarenheit`

File: Adafruit_MAX31855.cpp [hunk 1]
```diff
@@ -142,7 +328,7 @@ uint32_t Adafruit_MAX31855::spiread32(void) {
   if(sclk == -1) {
     // hardware SPI
 
-    SPI.beginTransaction(SPISettings(4000000, MSBFIRST, SPI_MODE0));
+    SPI.beginTransaction(SPISettings(4000000, MSBFIRST, SPI_MODE1));
```
Comment (tablatronix) [inline Adafruit_MAX31855.cpp]: This change is not covered in the PR

**Comment by caternuson** [comment]
Sorry for the delay in looking at this PR.

It looks like the main item is the updated computation for temperature following the NIST approach. The other items have been dealt with in other PRs. This PR would also require rebasing due to recent updates in the library. Since this new approach is a drop in replacement for the current temperature computation, it is OK to just replace the code in `readCelsisus` with this one.

Please consider resubmitting a PR with just those changes.

**Pull request closed without merging**